inline/src/parser/mod.rs

```diff
@@ -2,110 +2,181 @@ use std::{collections::VecDeque, ops::Deref};
+
+                    if cxt.merge_whitespace {
+                        // skip other whitespace
+                        while let Some(tkn) = self.next_token() {
+                            if !tkn.is_plain_whitespace() {
+                                self.token_cache = Some(tkn);
+                                break;
+                            }
+                        }
+                    }
+                } else {
+                    // cache popped token and break
+                    self.token_cache = Some(next_token);
+                    break;
+                }
+            } else {
+                let (next_content, next_span) = next_token.parts();
+                content.push_str(next_content);
+                span.end = next_span.end;
+            }
         }
 
-        while let Some(next_token) = self.next_token() {
-            end = next_token.span.end;
+        Inline::plain_or_eol(content, span, kind)
+    }
 
-            if next_token.closes() {
+    fn parse_nested(&mut self, start_token: Token) -> Inline {
+        let start = start_token.span.start;
+        let mut end = start_token.span.end;
+        let mut content = VecDeque::new();
+
+        while let Some(next_token) = self.next_token() {
+            if next_token.closes(Some(&start_token)) {
+                end = next_token.span.end;
                 break;
             } else if next_token.opens() {
                 // lexer resolved tokens, if token opens, it is guaranteed that closing exists too.
                 // If not, it's bug in implementation
                 let nested = self.parse_inline(next_token);
-                content.append_inline(nested);
+                content.push_back(nested);
             } else {
-                if kind == TokenKind::Plain && next_token.kind != TokenKind::Plain {
-                    self.token_cache = Some(next_token);
-                    break;
-                }
+                // opening token not yet closed, next token does not start a new inline -> parse as
+                // plain text
 
                 end = next_token.span.end;
-                content.append(InlineContent::from(next_token));
+                let (inner, span) = next_token.parts();
+                let inline = Inline::Plain(Plain {
+                    content: String::from(inner),
+                    span,
+                });
+                content.push_back(inline);
             }
         }
 
         let span = Span::from((start, end));
-        content.try_flatten();
-        Inline::with_span(content, kind, span)
+        Inline::nested_with_span(content, start_token.kind, span)
+    }
+
+    fn parse_inline(&mut self, token: Token) -> Inline {
+        // opening/closing of tokens is resolved at lexing stage
+        // at this point we can simply parse
+        let kind = token.kind;
+
+        let mut inline = match (kind, token.opens()) {
+            (TokenKind::Verbatim, _) => self.parse_plain(
```

Comment:
`Verbatim` is now also *kind-of* plain inline, because it just consumes any content found inside as a string. The context passed to `parse_plain` indicates what to do with consecutive whitespaces, which should handle the merging of whitespaces. 

We discussed internally how to represent multi-line strings so that we can also keep information about line-wise span information. That could potentially make it possible to handle multi-line verbatim inline in a better way (GitHub treats newlines in verbatim as a whitespace).

Reply:
GitHub follows the commonmark spec (https://spec.commonmark.org/0.30/#inlines), where `<code>` tags are used for inline verbatim, but `<pre><code>` for verbatim blocks. To still preserve non line-breaking whitespace in inline verbatim, they recommend adding CSS `code{white-space: pre-wrap;}`. 

I am still unsure if we should do the same, or use `<pre><code>` in both cases, but convert a line-break to space for inline verbatim. For now, I will create an issue in the spec repo.

You may skip this in your PR.